### DISCUSSION TITLE
nativepath for app: objects

### DISCUSSION THREAD

**Discussion by phitrti** [airsdk/Adobe-Runtime-Support#2622] [General]
I'm working on an app (currently android only) where I need to play a video. In the past, I've used the Distriqt MediaPlayer and either preloaded video content (12GB) or downloaded the video and saved it locally. In this case, it is a single instructional video, about 1:30 minutes long and I figured just bundle it as an asset along with everything else. When I ask for the nativepath of my video asset, "" is returned. Makes sense considering the design of "assets". No, I can not host the video on the web because the app needs to work offline or in a resource poor environment.

My question is, giving the recent work around File, is it possible to ask for the native path of these assets? According to the debugger, there are some private variables that "appear" to have native path information to this asset but there is no public way to get to them that I could find.

Currently, my lame work around is, when the app starts, to copy the video to applicationstorage, checking to see if it is there already. It isn't a big file but I hate having to have a duplicate. 

Randy

**Comment by ajwfrost** [comment]
With recent(ish) versions of Android, there may not be a native path for an `app:/` based URL - the assets are typically bundled into the APK and just left as zipped-up resources. On Android they are then accessed via a special API; with AIR, if you use an AS3 `File` object to get at them, then the AIR runtime will extract them to an area that's accessible under a normal Linux-style file path (all transparently to you, but it's happening under the hood so there's already a copy happening!)

So, are you saying that you have an asset as part of your APK and when you create a file via e.g. `File.applicationDirectory.resolvePath("myVideo.mp4")`, the resulting file has a blank `nativePath` value?

There may be scope for some optimisation, depending on how the video files are packaged up... it's possible to open an asset file as an `InputStream` where I would expect the Android subsystem then does the decompression and pulls the file out of the zip. Or if you know it's already uncompressed, you can use the below function to get at the file descriptor, which we could then use to provide the AS3 `File` capabilities...
https://developer.android.com/reference/android/content/res/AssetManager#openFd(java.lang.String)

I'm wondering what you need though -> is it a native path that is required so that you can pass this to another component (i.e. the ANE) or is it possible to provide an `app:/` style URL there? (As the ANE should then - theoretically at least - be able to translate the URI and use these Android functions to play back the video, I believe..)

**Reply by marchbold** [reply]
The issue here in the past was that assets were incorrectly compressed by AIR and hence couldn't be accessed with the `app://` url. However I believe you have resolved this issue and we just haven't looked into it again. I'll have a look into it again as I suspect this should work now.

**Reply by ajwfrost** [reply]
I am not sure whether our File API knows about assets being uncompressed or not, it will always just pull them out into a temporary location. But yes, good point in terms of the raw assets that had been compressed previously which means the `openFd` mechanism would fail.

But I think there's another bug (in AIR) that's preventing this currently, see below failure to resolve from File.applicationDirectory - we can check this one as it may have been affected by the recent changes to the file handling.

**Comment by phitrti** [comment]
> So, are you saying that you have an asset as part of your APK and when you create a file via e.g. File.applicationDirectory.resolvePath("myVideo.mp4"), the resulting file has a blank nativePath value?

Yes

> I'm wondering what you need though -> is it a native path that is required so that you can pass this to another component (i.e. the ANE) or is it possible to provide an app:/ style URL there? (As the ANE should then - theoretically at least - be able to translate the URI and use these Android functions to play back the video, I believe..)

Yes, an ANE is used for video playback and native path is required either local or web url. app:/ is not currently supported.

Thanks for the under the hood information. Will reach out to ANE developer about possibly adding app:/ as a future enhancement. For now, I have a working workaround. Not ideal but it works.

**Reply by ajwfrost** [reply]
Hi

Okay for this part:
> >So, are you saying that you have an asset as part of your APK and when you create a file via e.g. File.applicationDirectory.resolvePath("myVideo.mp4"), the resulting file has a blank nativePath value?
>
> Yes

I think this isn't new behaviour, the file object appears to always hide that path, as well as having a blank nativePath for the application directory as well (at least, trying with AIR 33.1 on an Android 12 handset as well as an older Android 9 one...)

An Android app/ane should be able to work from the "app:/" uri though, on the assumption that it knows a bit about AIR packaging (and assuming AIR/ADT hasn't compressed the required files...)

**Reply by marchbold** [reply]
Yeah the nativePath being empty is fine for packaged assets, we will need to use the url in these cases to get the content url `app://` etc.
